compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java

```diff
@@ -4923,6 +4943,321 @@ private Map<String, Location> getFieldLocations(List<BLangSimpleVariable> fields
+            Set<Flag> flagSet = classDefinition.flagSet;
+            if (!flagSet.contains(Flag.PUBLIC) || classDefinition.isServiceDecl || flagSet.contains(Flag.OBJECT_CTOR)) {
+                continue;
+            }
+
+            logErrorForUnsupportedConstructExposureFromClientDeclModule(packageID, collector, classDefinition.pos,
+                                                                        classDefinition.getBType());
+        }
+
+        for (BLangFunction function : bLangPackage.functions) {
+            if (!function.flagSet.contains(Flag.PUBLIC) &&
+                    (!function.attachedFunction || !function.receiver.flagSet.contains(Flag.PUBLIC))) {
+                continue;
+            }
+
+            logErrorForUnsupportedConstructExposureFromClientDeclModule(packageID, collector, function.pos,
+                                                                        function.getBType());
+        }
+    }
+
+    private void logErrorForUnsupportedConstructExposureFromClientDeclModule(PackageID currentPkgId,
+                         BClientDeclConstructExposureDisallowingTypeVisitor collector, Location location, BType type) {
+        List<BType> exposedTypes = new ArrayList<>(1);
+        collector.visitType(type, exposedTypes, currentPkgId);
+        if (!exposedTypes.isEmpty()) {
+            dlog.error(location,
+                       DiagnosticErrorCode.UNSUPPORTED_EXPOSURE_OF_CONSTRUCT_FROM_MODULE_GENERATED_FOR_CLIENT_DECL);
+        }
+    }
+
+    private static class BClientDeclConstructExposureDisallowingTypeVisitor implements TypeVisitor {
+
+        Set<BType> unresolvedTypes;
+        SymbolResolver symResolver;
+        SymbolTable symTable;
+        List<BType> exposedTypes;
+        PackageID currentPkgId;
+
+        public BClientDeclConstructExposureDisallowingTypeVisitor(SymbolResolver symResolver, SymbolTable symTable) {
+            this.symResolver = symResolver;
+            this.symTable = symTable;
+            this.unresolvedTypes = Collections.emptySet();
+            this.exposedTypes = Collections.emptyList();
+        }
+
+        public void visitType(BType type, List<BType> exposedTypes, PackageID currentPkgId) {
+            this.unresolvedTypes = new HashSet<>();
+            this.exposedTypes = exposedTypes;
+            this.currentPkgId = currentPkgId;
+            this.visitType(type);
+        }
+
+        private void visitType(BType type) {
+            if (type == null) {
+                return;
+            }
+
+            if (!unresolvedTypes.add(type)) {
+                return;
+            }
+
+            BTypeSymbol tsymbol = type.tsymbol;
+            if (tsymbol != null &&
+                    !this.currentPkgId.equals(tsymbol.pkgID) &&
+                    this.symTable.clientDeclarations.containsValue(tsymbol.pkgID)) {
+                this.exposedTypes.add(type);
+                return;
```

Comment:
Don't we need to remove `type` from `unresolvedTypes` before return in here.

Reply:
Why do we have to? `unresolvedTypes` is always a new set when the analysis is called.

Reply:
yes. agree with you now.